onnxruntime/core/session/inference_session.cc

```diff
@@ -1864,6 +1873,75 @@ AllocatorPtr InferenceSession::GetAllocator(const OrtMemoryInfo& mem_info) const
   return session_state_->GetAllocator(mem_info);
 }
 
+void InferenceSession::ShrinkDefaultMemoryArenas(const std::string& ort_device_list) {
+  if (ort_device_list.empty()) {
+    return;
+  }
+
+  std::stringstream ss_1(ort_device_list);
```

Comment:
Since this parsing is happening for every Run call, is there a faster way to achieve this? It might not matter for the small strings we're dealing with here, so might be premature but may show up for smaller models. Also, 'istringstream' instead of stringstream.

Reply:
Maybe the general 'shrink after EVERY Run' should be a session option (vs. a RunOption of 'shrink after THIS Run')

Reply:
However doing it as a RunOption has an interesting characteristic. The user could choose to only specify a shrink if they knew the input was unusually large.

Reply:
Keeping it as a RunOption for now - the string to be parsed on every run won't be too big and if need-be we can expose a session option to shrink on every run.